### PR TITLE
Add dynamic upload limits, configurable maxrambytes

### PR DESCRIPTION
**Pull request by EternityForest**
This pull request enables user code to dynamically accept or reject file uploads and large HTTP request parts with a callback function, before those files actually get written to disk.

It also allows configuring the threshold that determines if a given item is written to disk as a tempfile or just kept in RAM.

* **What is the related issue number (starting with `#`)**
#1498 

* **What is the current behavior?** (You can also link to an open issue here)
The temporary file threshold is set at 1000 bytes and​ cannot be changed in any documented way that I know of.

As far as I can tell, by default anyone can upload a large file which will be placed into a tempfile. As this happens before your application code, there's not much to be done to prevent unauthorized users from creating large tempfiles, using RAM or SSD write cycles.

* **What is the new behavior (if this is a feature change)?**
The request.body.maxbytes option may now be a tuple of (default, function). Up to default bytes will be read, after which function will be called with no arguments and whatever it returns will be used as the upload limit for that request. This allows your code to inspect HTTP headers to decide what to do.

The reason for the use of a default is to avoid calling the function in cases where it is a somewhat expensive operation, such as if it involves a call to a users database.

* **Other information**:
I believe this to be somewhat important for embedded systems, where RAM is limited and /tmp is often kept in a tmpfs to prevent wearing out the flash memory. In these situations it would be easy for an attacker to continually write to tmpfs until it was full, potentially using up all available tempfile space.

In addition, systems that do not use a tmpfs for /tmp could simply run out of disk space on small SD cards, or even wind up with totally destroyed flash on some very cheap cards without wear leveling.

### CODE REVIEW

File: cherrypy/_cprequest.py [hunk 1]
```diff
@@ -643,7 +643,7 @@ def respond(self, path_info):
 
                     self.body = _cpreqbody.RequestBody(
                         self.rfile, self.headers, request_params=self.params)
-
+
```
Comment (webknjaz) [inline cherrypy/_cprequest.py]: Changing this file is absolutely unnecessary

File: docs/advanced.rst [hunk 2]
```diff
@@ -447,7 +447,8 @@ and wrap your entire CherryPy application with it:
    def secureheaders():
        headers = cherrypy.response.headers
        headers['X-Frame-Options'] = 'DENY'
-       headers['X-XSS-Protection'] = '1; mode=block'
+       headers['        self.assertStatus(413)
```
Comment (webknjaz) [inline docs/advanced.rst]: This looks accidentally changed

**Review by webknjaz — CHANGES_REQUESTED** [review]
There's a bunch of linters failed: https://travis-ci.org/cherrypy/cherrypy/jobs/246086738

You may use `pre-commit` tool to check it locally.
Running the following must help you:
```shell
$ pre-commit run --all-files
```

File: cherrypy/_cpreqbody.py [hunk 3]
```diff
@@ -724,14 +738,22 @@ def default_proc(self):
...
 
         Return fp_out.
         """
-        if fp_out is None:
-            fp_out = self.make_file()
-        self.read_lines_to_boundary(fp_out=fp_out)
-        return fp_out
+        if not fp_out:
+            x = self.read_lines_to_boundary()
+            if isinstance(x,(str,bytes)):
```
Comment (webknjaz) [inline cherrypy/_cpreqbody.py]: I'm not sure whether this block would work under both Python 2 and Python 3. There's [`_cpcompat` module having helper functions](https://github.com/cherrypy/cherrypy/blob/1639205/cherrypy/_cpcompat.py#L27-L91) for dealing with different string/byte representations.

File: cherrypy/_cpreqbody.py [hunk 4]
```diff
@@ -724,14 +738,22 @@ def default_proc(self):
...
+        a BytesIO object, based on maxrambytes.
 
         Return fp_out.
         """
-        if fp_out is None:
-            fp_out = self.make_file()
-        self.read_lines_to_boundary(fp_out=fp_out)
-        return fp_out
+        if not fp_out:
+            x = self.read_lines_to_boundary()
```
Comment (webknjaz) [inline cherrypy/_cpreqbody.py:747]: You can do this call once before the loop with `fp_out`.

File: cherrypy/_cpreqbody.py [hunk 5]
```diff
@@ -802,7 +843,13 @@ def read(self, size=None, fp_out=None):
             # Check lengths.
             self.bytes_read += datalen
             if self.maxbytes and self.bytes_read > self.maxbytes:
-                raise cherrypy.HTTPError(413)
+                if self.maxbytes_extend:
+                    print("extending maxbytes")
```
Comment (webknjaz) [inline cherrypy/_cpreqbody.py]: This statement is not needed. If you want to log smth, use logging

File: cherrypy/_cpreqbody.py [hunk 6]
```diff
@@ -802,7 +843,13 @@ def read(self, size=None, fp_out=None):
...
+                    self.maxbytes = self.maxbytes_extend()
+                    #We only want to call maxbytes extend once per request I think
+                    self.maxbytes_extend =None
+                if self.maxbytes and self.bytes_read > self.maxbytes:
```
Comment (webknjaz) [inline cherrypy/_cpreqbody.py]: Do `bytes_read` and `maxbytes` mutate since previous check? If not, this conditional doesn't make any sense, because it's always `True` at this point.

File: cherrypy/_cpreqbody.py (hunk 3 above)
Comment (EternityForest) [reply cherrypy/_cpreqbody.py]: Seems to work on both in the interactive shell, bytes appears to be an alias of string in python. There's probably a better way, but this seemed like a decent way to express "Is this value object and not a file like object?"

Comment (webknjaz) [reply cherrypy/_cpreqbody.py]: `str` is alias of unicode type in Python 3 and bytes in Python 2. I'd go for try/except following [EAFP principle](https://docs.python.org/3/glossary.html).

File: cherrypy/_cpreqbody.py (hunk 5 above)
Comment (EternityForest) [reply cherrypy/_cpreqbody.py]: Uh oh. That was a debug print statement that I was using(So I could do a search for print statements and delete it easier than searching through the log calls).

Looks like I forgot the last step.

File: cherrypy/_cpreqbody.py (hunk 3 above)
Comment (webknjaz) [reply cherrypy/_cpreqbody.py]: You could also use `isinstance(x, file)`

Comment (EternityForest) [reply cherrypy/_cpreqbody.py]: So just check if the BytesIO constructor fails due to getting a file-like object? I like the explicitness of checking for the specific type, but it does kind of make it look like x might be an actual unicode string if you don't know the python2 history.

Either way there should probably be a comment there I'd say.

Comment (webknjaz) [reply cherrypy/_cpreqbody.py]: Two options are:
1) `isinstance(x, file)`
2) try/`except TypeError`

**Comment by EternityForest** [comment]
What's the policy for amending commits here? Amend, squash, add a new commit?

**Comment by webknjaz** [comment]
@EternityForest just maintain a clean history. Add commits if needed. You may want to do a little history cleanup in the end of review. Anyway GitHub has 3 ways of merging and if we want to squash, we'll just hit appropriate button in PR :)

File: cherrypy/_cpreqbody.py [hunk 7]
```diff
@@ -582,6 +581,7 @@ def process(self):
             proc(self)
 
     def default_proc(self):
+
```
Comment (webknjaz) [inline cherrypy/_cpreqbody.py:584]: Why did you add this line?

File: cherrypy/_cpreqbody.py [hunk 8]
```diff
@@ -724,14 +736,22 @@ def default_proc(self):
...
-            fp_out = self.make_file()
-        self.read_lines_to_boundary(fp_out=fp_out)
-        return fp_out
+        if not fp_out:
+            x = self.read_lines_to_boundary()
+            try:
+                return io.BytesIO(x)
+            except TypeError:
+                return x
+        else:
```
Comment (webknjaz) [inline cherrypy/_cpreqbody.py:752]: I'd remove `else`-block to reduce code nesting

File: cherrypy/_cpreqbody.py (hunk 4 above)
Comment (webknjaz) [reply cherrypy/_cpreqbody.py:747]: I still think this function should be called once before the `if`-block

Comment (webknjaz) [reply cherrypy/_cpreqbody.py:747]: Also, please use meaningful variable name instead of `x`.

File: cherrypy/_cpreqbody.py (hunk 7 above)
Comment (webknjaz) [reply cherrypy/_cpreqbody.py:584]: Try doing as little unrelated changes as possible

File: cherrypy/_cpreqbody.py [hunk 9]
```diff
@@ -724,14 +736,22 @@ def default_proc(self):
                 self.file = result
 
     def read_into_file(self, fp_out=None):
-        """Read the request body into fp_out (or make_file() if None).
+        """Read the request body into fp_out.
+
+        if fp_out not given, use either make_file or
```
Comment (webknjaz) [inline cherrypy/_cpreqbody.py:741]: Please start sentence with capital letter

File: cherrypy/_cpreqbody.py [hunk 10]
```diff
@@ -214,6 +215,9 @@ def process_multipart(entity):
     # Read all parts
     while True:
         part = entity.part_class.from_fp(entity.fp, ib)
+        # Inherit maxrambytes from the parent entity.
```
Comment (webknjaz) [inline cherrypy/_cpreqbody.py:218]: Please move this to `Part.from_fp` classmethod

File: cherrypy/_cpreqbody.py [hunk 11]
```diff
@@ -745,7 +765,13 @@ def __init__(self, fp, length, maxbytes, bufsize=DEFAULT_BUFFER_SIZE,
         # Wrap our fp in a buffer so peek() works
         self.fp = fp
         self.length = length
-        self.maxbytes = maxbytes
+        if isinstance(maxbytes, tuple):
```
Comment (webknjaz) [inline cherrypy/_cpreqbody.py:768]: This should work for any iterable, I think. Not just tuples.

File: cherrypy/_cpreqbody.py [hunk 12]
```diff
@@ -832,8 +876,12 @@ def read(self, size=None, fp_out=None):
             # Check lengths.
             self.bytes_read += datalen
             if self.maxbytes and self.bytes_read > self.maxbytes:
-                raise cherrypy.HTTPError(413)
-
+                if self.maxbytes_extend:
```
Comment (webknjaz) [inline cherrypy/_cpreqbody.py:879]: This block of code duplicates code above. It must be moved to separate private/protected method

**Review by webknjaz — CHANGES_REQUESTED** [review]
Please also add docstrings to functions you added

File: cherrypy/_cpreqbody.py (hunk 11 above)
Comment (webknjaz) [inline cherrypy/_cpreqbody.py:768]: After thinking this through, I recall that there's some places in config, where we allow setting callable instead of scalar as a value. We should probably allow setting `maxbytes` to callable and test it like:
```python
try:
    self.maxbytes = maxbytes()
except TypeError:
    self.maxbytes = maxbytes
```

Comment (webknjaz) [reply cherrypy/_cpreqbody.py:768]: @jaraco any additional thoughts?

**Comment by webknjaz** [comment]
Closing this, since there's three solutions provided within #1498 + this one is incomplete and I don't have time to work on it myself.
